### PR TITLE
fix blas-lapack in scipy and numpy

### PR DESCRIPTION
**Pull request by davydden**
macOS Sierra: 

- [x] `spack install py-scipy ^openblas`

p.s. can't test with `atlas` as it is broken on Sierra.

Ubuntu 16.04+gcc@5.4.0 with uninstalled system blas/lapack:

- [x] `spack install py-scipy ^openblas`
- [x] `spack install py-scipy ^atlas`

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-numpy/package.py [hunk 1]
```diff
@@ -66,21 +66,18 @@ def setup_dependent_package(self, module, dep_spec):
...
+        if '+lapack' in spec:
+            lapackblas = lapackblas + spec['lapack'].lapack_libs
 
         if '+blas' in spec:
-            libraries.append('blas')
-            library_dirs.append(spec['blas'].prefix.lib)
-        if '+lapack' in spec:
-            libraries.append('lapack')
-            library_dirs.append(spec['lapack'].prefix.lib)
+            lapackblas = lapackblas + spec['blas'].blas_libs
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-numpy/package.py]: Python has a `+=` operator.

**Comment by davydden** [comment]
`py-numpy` site.cfg is now
```
[DEFAULT]
libraries=openblas
library_dirs=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/openblas-0.2.19-7qdxwznmqzlbqtusssagtxkypivod6ct/lib
rpath=/Users/davydden/spack/opt/spack/darwin-sierra-x86_64/clang-8.0.0-apple/openblas-0.2.19-7qdxwznmqzlbqtusssagtxkypivod6ct/lib
```

**Comment by davydden** [comment]
[documentation](http://www.scipy.org/scipylib/building/linux.html) of `py-scipy` is a bit contradicting: 

in some places they say that you need to 
```
export BLAS=/path/to/libblas.so
export LAPACK=/path/to/liblapack.so
```
In others, they just rely on what is used during `py-numpy` build.

**Comment by citibeth** [comment]
I'm frankly dubious about this strategy. The previous approach was based on the idea that a virtual dependency (such as `blas`) should present a consistent API to the world.  In this case, the proposed API was a library called `libblas.so` or `libblas.a`.  Implementations produce files of their own names, of course, but should link to the standard (API) names.  I don't see how this is such a bad approach, or why it wouldn't work with ATLAS.

The API that implementations of BLAS provide a Python variable called `blas_libs` is a good one, of course... except that `py-scipy` has hard-coded somewhere inside that the library is named `libblas.so`.  The symlinks do the same thing as the Python API, but a bit more forcefully.  If it were implemented consistently across all implementations of BLAS, it would not bother me.

The strategy in this PR feels, quite frankly, like a hack.  Why should `py-numpy` be stashing away the location of BLAS for `py-scipy` to pick up?  And why is it better to stash that information away in a file we just invented in `py-numpy`'s install dir, rather than in `openblas` install dir?  The original problem was that `py-scipy` couldn't find BLAS, it did not include `py-numpy`. 

It just seems more complicated than having BLAS implementations tell the world about themselves.  This PR is also less general --- we'll have to do a similar fix-up the next time we run across a brain-dead package that assumes BLAS is named `libblas.so`.

I vote for continued discussion on this issue to come to more of a consensus.

**Comment by citibeth** [comment]
As for testing... if this PR is to be merged, please test with `spack install py-scipy ^openblas` before merging.

**Comment by davydden** [comment]
> The previous approach was based on the idea that a virtual dependency (such as blas) should present a consistent API to the world. In this case, the proposed API was a library called libblas.so or libblas.a

that is not possible. Look at the list of libs for atlas and MKL, none of those are a single lib or a pair of `blas`, `lapack`. 

It would not work with Atlas, because it has several libs for static case:
```
if '+shared' in self.spec:
            to_find = ['libtatlas'] if is_threaded else ['libsatlas']
            shared = True
        else:
            interfaces = [
                'libptcblas',
                'libptf77blas'
            ] if is_threaded else [
                'libcblas',
                'libf77blas'
            ]
            to_find = ['liblapack'] + interfaces + ['libatlas']
            shared = False
        return find_libraries(
            to_find, root=self.prefix, shared=shared, recurse=True
        )

```

 It would not work with MKL becase it has several libs for dynamic case:
```
shared = True if '+shared' in self.spec else False
        suffix = dso_suffix if '+shared' in self.spec else 'a'
        mkl_integer = ['libmkl_intel_ilp64'] if '+ilp64' in self.spec else ['libmkl_intel_lp64']  # NOQA: ignore=E501
        mkl_threading = ['libmkl_sequential']
        if '+openmp' in self.spec:
            mkl_threading = ['libmkl_intel_thread', 'libiomp5'] if '%intel' in self.spec else ['libmkl_gnu_thread']  # NOQA: ignore=E501
        # TODO: TBB threading: ['libmkl_tbb_thread', 'libtbb', 'libstdc++']
        mkl_libs = find_libraries(
            mkl_integer + ['libmkl_core'] + mkl_threading,
            root=join_path(self.prefix.lib, 'intel64'),
            shared=shared
        )
        system_libs = [
            'libpthread.{0}'.format(suffix),
            'libm.{0}'.format(suffix),
            'libdl.{0}'.format(suffix)
        ]
        return mkl_libs + system_libs
```


> The strategy in this PR feels, quite frankly, like a hack. Why should py-numpy be stashing away the location of BLAS for py-scipy to pick up? 

Please, reed the link http://www.scipy.org/scipylib/building/linux.html#step-4-build-numpy-1-5-0 This is my interpretation of how developers designed the config system.

> The original problem was that py-scipy couldn't find BLAS, it did not include py-numpy.

correct, that's because `py-numpy` was happy to build with anything, then its config was picked up in `py-scipy`. Again, this is my interpretation of what is happening.

> It just seems more complicated than having BLAS implementations tell the world about themselves. 

It is, but this is the reality of blas/lapack packages.


> please test with spack install py-scipy ^openblas before merging.

see the OP.

**Comment by tgamblin** [comment]
@citibeth: The idea of this strategy IS for packages to tell the world about themselves, but for them to do that within Spack.  I do not want to rely on a directory structure/file naming strategy that we create, when the package ships from the developer using a different strategy.  There are a few reasons for this:

1. BLAS is a nightmare and the implementations (nearly) all present different library names. I think SciPy is doing it wrong by assuming libels.[so|a].  You could argue that BLAS *should* do something better, but given that some implementations don't even come with headers (sigh... Fortran) I'm not optimistic that'll happen.

1. We may not be able to take all BLAS implementations and condense them to a single library symlink.  Some come with multiple libraries.  The proposed interface allows lists of libraries to address that.

1. Relying on symlinks in Spack makes the way external deps (which will have a native layout) and internal deps (where we could symlink things) inconsistent.  The `blas_libs` vars are designed to give builds the information they need to link with _either_ a spack-built package or an external.  I don't think we can guarantee that will work with symlinks.

So the basic idea from @davydden and @alalazo put together allows us to have an inter-spec interface in Spack.  The nice thing about this is that it can be used with other packages that expose inconsistent library APIs.  I agree with you that library naming should really be PART of a decent API, but there are a lot of libraries for which it's not (MPI, BLAS, LAPACK, others).  MPI gets around it with compiler wrappers and the standard punts on any link-level API.  BLAS and LAPACK are informal and at best source-level interfaces... Spack needs to work with that stuff and I think the spec interface handles it better than symlinking, especially where different versions of a vdep may expose different names and numbers of libraries.

**Comment by citibeth** [comment]
Thanks @davydden you've convinced me.  Yet another upstream package with a crazy build procedure...

**Comment by tgamblin** [comment]
I think there are lots of good visual metaphors for upstream in HPC.

kayak
bear and salmon

**Comment by tgamblin** [comment]
So more importantly -- does this work for you, @citibeth?

**Comment by citibeth** [comment]
As long as `py-scipy` builds, it works for me.  Since my problem was
(probably) not related to my OS or configuration, I'll assume it works if
someone else can build it.

**Comment by davydden** [comment]
will test it on Ubuntu now...

**Comment by davydden** [comment]
Successfully built `py-scipy` with both `atlas` and `openblas` on Ubuntu, so there should be no degradation of blas/lapack support due to this PR. Ready to be merged, as far as I am concerned.

**Comment by citibeth** [comment]
We've collectively spent a lot of time on this issue.  Some documentation improvements would be in order to record what we've learned here:

BLAS/LAPACK API
---------------

1. I like the design of blas_libs and lapack_libs lists on the specs.  I recommend a section explaining this gets added to the docs, and then reference that section in every package that implements `blas` or `lapack` (one-line comment).

site.cfg
--------

The `site.cfg` thing is... weird.  I recommend comments in the Spack code explaining:

 1. This is Numpy/Scipy thing, not Spack

 2. Where to find more information about it on-line (some of the URLs
    referenced above)

 3. Maybe a link to this tread?

These comments would go in the packages for `py-numpy` and `py-scipy`.

Code nitpick
-------------

For lapackblaslibs, could some kind of set be used, to avoid duplicates between blas and lapack?  Or maybe better... is there a method on LibraryList() that can be used to eliminate duplicates at the end?

**Comment by davydden** [comment]
> I recommend a section explaining this gets added to the docs

There is something, see https://github.com/LLNL/spack/pull/1682/files#diff-4b5b30a2f8dadfd4a7962d62cac8838aL2091

> The site.cfg thing is... weird. I recommend comments in the Spack code explaining:

already there, see https://github.com/LLNL/spack/pull/1949/files#diff-dec961e02b96009662438a9452626339R53

>  is there a method on LibraryList() that can be used to eliminate duplicates at the end?

already there, see https://github.com/LLNL/spack/pull/1682/files#diff-7b0a939725679145b7e09f1104456b61R377

**Comment by citibeth** [comment]
LGTM

**Pull request merged**